### PR TITLE
flyspell-correct-completing-read: Improve shortcut key

### PR DESCRIPTION
**Pull request by minad**
The RET is pressed automatically for you.

(Follow up on #92 - I used it for a while and figured that pressing the RET after the actions is a bit tedious.)

### CODE REVIEW

**Comment by minad** [comment]
Converted to draft, I am still refining.

**Comment by manuel-uberti** [comment]
Could the automation be made opt-in? I like the shortcuts but I wouldn't want a wrong option to be automatically picked, where "wrong" means I meant to type something else but I did it quickly without thinking too much about it.

**Comment by minad** [comment]
> Could the automation be made opt-in? I like the shortcuts but I wouldn't want a wrong option to be automatically picked, where "wrong" means I meant to type something else but I did it quickly without thinking too much about it.

One could but I think there is actually no reason. It will be made such that the quick selection is always safe. Either you press @ for an action or a digit for a suggestion. Both keys (@ and digits) are hard to press accidentally.

**Comment by manuel-uberti** [comment]
I agree, you're right about `@` not being that quick to type. :)

**Comment by minad** [comment]
Okay, it is ready now. @manuel-uberti If you have time, please give it a try. The quick keys work both for Ido and completing-read. I tested default completion, Vertico and Selectrum.

**Comment by manuel-uberti** [comment]
I tried it with Vertico, and `@` plus the quick key worked without having to press `RET`. :+1:

File: flyspell-correct.el [hunk 1]
```diff
@@ -111,17 +111,26 @@ highlighting."
...
+  '((save ?s "[Save]")
+    (session ?a "[Accept (session)]")
+    (buffer ?b "[Accept (buffer)]")
+    (skip ?k "[Skip]")
+    (stop ?p "[Stop]"))
+  "Actions used by `flyspell-correct-completing-read'.")
+
+(defun flyspell-correct--cr-index (n)
+  "Generate a short unique index string for N."
+  (let ((str (char-to-string (+ ?1 (mod n 5)))))
```
Comment (d12frosted) [inline flyspell-correct.el:127]: Why `5`? I would rather use length of `flyspell-correct--cr-actions` - more future proof.

**Comment by d12frosted** [comment]
Oh, I like this idea. Good job!

File: flyspell-correct.el (hunk 1 above)
Comment (minad) [reply flyspell-correct.el:127]: The `flyspell-correct--cr-index` is used for the indexes prepended before the suggestions. Base 5 is used here to result in nice index strings. The indices are generated such that they terminate with 1,2,3,4,5 such that we can commit if one of these keys are pressed.

~~~ elisp
(mapcar #'flyspell-correct--cr-index (number-sequence 0 100))

("1" "2" "3" "4" "5" "61" "62" "63" "64" "65" "71" "72" "73" "74" "75" "81" "82" "83" "84" "85" "91" "92" "93" "94" "95" "01" "02" "03" "04" "05" "661" "662" "663" "664" "665" "671" "672" "673" "674" "675" "681" "682" "683" "684" "685" "691" "692" "693" "694" "695" "601" "602" "603" "604" "605" "761" "762" "763" "764" "765" "771" "772" "773" "774" "775" "781" "782" "783" "784" "785" "791" "792" "793" "794" "795" "701" "702" "703" "704" "705" "861" "862" "863" "864" "865" "871" "872" "873" "874" "875" "881" "882" "883" "884" "885" "891" "892" "893" "894" "895" "801")
~~~

Comment (d12frosted) [reply flyspell-correct.el:127]: Ah, I see. Misunderstood the intent here. Thanks for explanation.

**Comment by minad** [comment]
@d12frosted Which frontend do you use btw? With this PR I think the CR frontend is almost as convenient as the avy-menu/frog-menu frontends, with the additional possibility to scroll/search through the items.

**Comment by d12frosted** [comment]
@minad I am using Selectrum. For a long time I've been using ivy over helm for simplicity™. But over time it become terrifically complex, so I wanted something simpler for completion, closer to default frontend, but vertical. So... Here I am :) 

Now regarding your PR. I find it convenient and it boosts my velocity even better than `flyspell-correct-ivy` interface. Seriously, bravo @minad . 

I am ready to merge this, but this needs to be documented. (1) I would love `flyspell-correct--cr-index` to contain your explanation in the docstring and (2) README 😸 If you don't feel like documenting - that's fine, I will do it, just let me know.

And one point that is out of scope of this PR, but I still would love to hear your opinion. What do you think about making it possible to configure indexing? While I find `flyspell-correct--cr-index` to be clever ergonomics wise, it's still an ergonomics hack for certain context, and maybe others would love to change it to natural numbers or something else.

**Comment by minad** [comment]
> Now regarding your PR. I find it convenient and it boosts my velocity even better than flyspell-correct-ivy interface. Seriously, bravo @minad.

Thanks! That's good to hear.

> I am ready to merge this, but this needs to be documented. (1) I would love flyspell-correct--cr-index to contain your explanation in the docstring and (2) README

I will improve the documentation and then you can take another look at and maybe improve if further according to your preferences.

> And one point that is out of scope of this PR, but I still would love to hear your opinion. What do you think about making it possible to configure indexing? While I find flyspell-correct--cr-index to be clever ergonomics wise, it's still an ergonomics hack for certain context, and maybe others would love to change it to natural numbers or something else.

Yes, I think this is out of scope. Note that it is not easily possible to change this to natural numbers, since then we cannot auto submit. The idea is that the last digit is 1,2,3,4,5 which submits. The digits 6,7,8,9,0 are used to encode the number/5. This is not easy to change. I thought a bit about a nice scheme for the quick keys and this is what I came up with. I am sure someone users have their special preferences, but then they can make another proposal or even make a PR.

**Comment by d12frosted** [comment]
> Yes, I think this is out of scope. Note that it is not easily possible to change this to natural numbers, since then we cannot auto submit. The idea is that the last digit is 1,2,3,4,5 which submits. The digits 6,7,8,9,0 are used to encode the number/5. This is not easy to change. I thought a bit about a nice scheme for the quick keys and this is what I came up with. I am sure someone users have their special preferences, but then they can make another proposal or even make a PR.

Right. Automatic commit breaks then. 

> I will improve the documentation and then you can take another look at and maybe improve if further according to your preferences.

Do you wish to make a separate PR for documentation changes?

**Comment by minad** [comment]
> Do you wish to make a separate PR for documentation changes?

Just pushed it here, I hope it is good like this.

File: README.org [hunk 2]
```diff
@@ -117,8 +117,13 @@ If you do not want any binding, just replace =:bind (:map
...
 to use lazy loading.
 
-In order to select the (s)ave, (a)ccept, sto(p) or s(k)ip actions, you can
-enter the shortcut key @s, @a, @p or @k.
+In order to select the (s)ave, (a)ccept, sto(p) or s(k)ip actions, you can enter
+the shortcut key =@s=, =@a=, =@p= or =@k=. The actions will be automatically submitted.
+Furthermore suggestions can be quickly submitted by entering the numeric index
+in front of the suggestion. Besides the quick keys the usual ~completing-read~
+interface is available, which may be enhanced and allow scrolling through
+candidates if you have a completion UI like Icomplete-vertical, Selectrum or
```
Comment (d12frosted) [inline README.org:125]: Please add links to these projects 😸

**Review by d12frosted — APPROVED** [review]

File: README.org (hunk 2 above)
Comment (minad) [reply README.org]: Done

**Comment by d12frosted** [comment]
Thank you very much, merged as a2c5dbf, 5cb53ac and b3b7b72 💯 🎉

**Pull request closed without merging**